### PR TITLE
clean up sync in async usage

### PR DESCRIPTION
**Pull request by jacobheun**
This cleans up some logic in the identify flow where sync calls were being done in async code, which can lead to stack bloating. This moves the sync logic out of the async control flow.

Relates to https://github.com/libp2p/js-libp2p-switch/issues/287

### CODE REVIEW

File: src/connection/manager.js [hunk 1]
```diff
@@ -145,20 +145,18 @@ class ConnectionManager {
...
+            ], (err, peerInfo, observedAddrs) => {
               if (err) {
                 return muxedConn.end(() => {
                   callback(err, null)
                 })
               }
 
+              observedAddrs.forEach((oa) => {
+                this.switch._peerInfo.multiaddrs.addSafe(oa)
+              })
```
Comment (mcollina) [inline src/connection/manager.js:152]: I would recommend using a for loop here instead

File: src/connection/manager.js [hunk 2]
```diff
@@ -145,20 +145,18 @@ class ConnectionManager {
...
-              (conn, cb) => identify.dialer(conn, cryptoPI, cb),
-              (peerInfo, observedAddrs, cb) => {
-                observedAddrs.forEach((oa) => {
-                  this.switch._peerInfo.multiaddrs.addSafe(oa)
-                })
-                cb(null, peerInfo)
-              }
-            ], (err, peerInfo) => {
+              (conn, cb) => identify.dialer(conn, cryptoPI, cb)
+            ], (err, peerInfo, observedAddrs) => {
```
Comment (mcollina) [inline src/connection/manager.js]: You might want to remove that waterfall call, it would improve things as well.

**Comment by jacobheun** [comment]
I refactored the waterfall logic to use awaits and created some utility methods to make that cleaner. Also removed `forEach` in favor of `for(;;)`

**Review by vasco-santos — APPROVED** [review]
LGTM!

**Comment by dirkmc** [comment]
With regards to switching forEach loops to for loops etc, I would recommend instead focusing performance improvements on changes that make a difference in terms of big O notation: https://en.wikipedia.org/wiki/Big_O_notation

File: src/connection/manager.js [hunk 3]
```diff
@@ -136,34 +136,33 @@ class ConnectionManager {
...
+              results = await identifyDialer(msConn, cryptoPI)
+            } catch (err) {
+              return muxedConn.end(() => {
+                callback(err, null)
+              })
+            }
+
+            const { peerInfo, observedAddrs } = results
+
+            for (var i = 0; i < observedAddrs.length; i++) {
```
Comment (dirkmc) [inline src/connection/manager.js:157]: Can you use a `for ... of` loop here instead?

File: src/utils.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
+'use strict'
```
Comment (dirkmc) [inline src/utils.js:1]: Can you use promisify instead of these functions? https://github.com/digitaldesignlabs/es6-promisify#readme

Comment (jacobheun) [reply src/utils.js:1]: I added these as an interim solution. We're going to be doing a lot of work soon for they async iterators initiative https://github.com/libp2p/js-libp2p/issues/266 which will remove the need for these or promisfy. If that ends up taking longer than we'd like, or we have more immediate needs in other parts of the code base, then I think switching over to promisify is warranted.

Comment (dirkmc) [reply src/utils.js:1]: 👍

File: src/connection/manager.js (hunk 3 above)
Comment (mcollina) [reply src/connection/manager.js:157]: `for ... of ` suffers from some performance issues still.

**Comment by jacobheun** [comment]
In regards to overall performance improvements, we have an OKR for Q1 to get benchmarking in place for libp2p. That along with the current work for ipfs benchmarking will put us in a better position to do isolated analyses to see where the bottlenecks are in switch and the rest of libp2p. While I think those are very important, I want to avoid pushing more of that into this particular PR, as it's goal was to fix an issue with synchronous code being executed asynchronously, to help alleviate #287. I appreciate all the great feedback!

**Comment by mcollina** [comment]
> With regards to switching forEach loops to for loops etc, I would recommend instead focusing performance improvements on changes that make a difference in terms of big O notation: https://en.wikipedia.org/wiki/Big_O_notation

I agree, the main gains are in algorithmic improvements in big O notation. However, there is a lot to gain from making the code more optimizable by V8.

As an example, `forEach` would do  `n` function calls, which would not appear in a big O analysis. These have a cost in term of memory allocation and work inside V8, that would not bubble up when just reasoning about algorithms. Each one of these can easily eat 5-10% of the time of a benchmark in a hot path.

This is not always a hot path, but I've see this function in a significant number of flamegraphs.

**Pull request merged**